src/Microsoft.Extensions.Configuration.AzureAppConfiguration/DnsClient/DnsProcessor.cs

```
@@ -0,0 +1,390 @@
+
+        public async Task<IReadOnlyCollection<SrvRecord>> QueryAsync(IPEndPoint endpoint, string query, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var udpClient = new UdpClient(endpoint.AddressFamily);
+
+            var resultRecords = new List<SrvRecord>();
+
+            try
+            {
+                using var callback = cancellationToken.Register(() =>
+                {
+#if !NET45
+                    udpClient.Dispose();
+#else
+                    udpClient.Close();
+#endif
+                });
+
+                var originRecords = await QueryAsyncInternal(endpoint, $"{OriginSrvPrefix}.{query}", udpClient, cancellationToken).ConfigureAwait(false);
+                string originHost = query;
+                if (originRecords != null && originRecords.Count > 0)
+                {
+                    originHost = originRecords.First().Target;
+                }
+                else
+                {
+                    // If can't get any records from _origin query, we should return;
+                    return resultRecords;
+                }
+
+                IReadOnlyCollection<SrvRecord> altRecords = null;
+                ushort index = 0;
+
+                while (true)
+                {
+                    altRecords = await QueryAsyncInternal(endpoint, $"{AlternativeSrvPrefix(index)}.{originHost}", udpClient, cancellationToken).ConfigureAwait(false);
+
+                    if (altRecords == null || altRecords.Count == 0)
+                    {
+                        break;
+                    }
+
+                    resultRecords.AddRange(altRecords);
+
+                    index++;
+                }
+
+                return resultRecords;
+            }
+            catch (SocketException se) when (se.SocketErrorCode == SocketError.OperationAborted)
+            {
+                throw new TimeoutException();

```

Seems like we should be catching these thrown exceptions like `TimeoutException` and `DnsResponseException` in `TryRefreshAsync` and wherever else we might be calling this

---

`TryRefreshAsync` would not call this method directly, it calls the `SrvLookupClient.QueryAsync()` method, it has catch all exceptions and logging warning, does it make sense to you?

---

My fault, I didn't see that there were 2 `QueryAsync` methods. Makes sense!